util/dvsim/SimCfg.py

```diff
@@ -78,6 +78,39 @@ def resolve_dump_format(tool, dump):
     return pick_dump_format(fmts)
 
 
+def expand_run_list(tests, build_map, cfg):
+    '''Generate a list of tests to be run
+
+    For each test in tests, we add it test.reseed times. The ordering is
+    interleaved so that we run through all of the tests as soon as possible.
+
+    If build_map is not None, add each test to it (signifying that the test
+    should be built once the build on which it depends is done).
+
+    cfg should be a SimCfg object, passed to the RunTest constructor.
+
+    '''
+    tagged = []
+    for test in tests:
+        for idx in range(test.reseed):
+            tagged.append((idx / test.reseed,
```

Comment:
```suggestion
            tagged.append((idx,
```
This fixes the issue I pointed out in the image I shared in an earlier comment.

Reply:
As I wrote more than once in the attached issue, one has to decide what should happen if one test has more runs than another. If A should run 2 times and B should run 4, you have to decide whether to run ABABBB or ABBABB. Your proposed code runs ABABBB. My current code (I hope) runs ABBABB, which was intentional!

In the screenshot you posted, you see the `uart_fifo_reset` test running more than the others. That's because they have reseed = 50 and it has reseed = 300. I think this is exactly what is supposed to happen.

Note that if you decide to interleave things "fairly" (which I what my code tries to do) you also have to decide on whether to order by the start or the middle of the intervals. That is, if test A runs twice, then you can imagine the two runs splitting the interval [0, 1] into [0, 0.5], [0.5, 1]. Do we order by the starts of those intervals (0 and 0.5) or their middles (0.25, 0.75)? Choosing the starts as I do in this code (and then using a stable sort) ensures that the first things that run will always be one iteration of each test, in the order given by the hjson file.

Reply:
I think at times one would also like to know if all iterations of one specific tests passed to get an early feedback of how our updates affected it. For this, it is better to have a more predictable dispatch rather than the fair dispatch. Taking my earlier example again (A-5, B-10, C-2), if I am interested in knowing if all iterations of C passed, I just need to wait for 6 tests to dispatch and retire. With the fair dispatch, it is not as easy to predict when will all tests of C be done, and it is a longer wait (from the example you indicated in your comment - `ABCBABBABCBABBABB`).

Reply:
I'm not sure it makes sense to talk about doing "all the runs" for a test. If we don't fix seeds, we're always just doing a soak test with some arbitrary cut-off. I would have thought that you want to spend a roughly constant fraction of your time on each test. With your example, if you believe that testing C more would be helpful, you should probably change the weighting!

Do you feel strongly that the current code is wrong? If so, I'll change it to your preferred version. Otherwise, I think the current code is better.

Reply:
> I'm not sure it makes sense to talk about doing "all the iterations" for a test. If we don't fix seeds, we're always just doing a soak test with some arbitrary cut-off. 

Not sure what you mean here. The # of iterations set for each test is driven by coverage (and to some extent the MTBF); so each test is potentially reseeded different number of times.  

My point is lets say we were running 20 iterations of foo test and 1 of the iterations is failing consistently in the nightly regressions. We debugged and fixed the code and reran the whole regression locally. Isn't it useful to get an early feedback that all 20 iterations of foo test are now passing (and in the process also confirm that our fix did not break any other test) than wait all the way till the end of the regression?

I would have thought that you want to spend a roughly constant fraction of your time on each test, rather than running one test lots at the start and then not at all later. 

With your fix, are we not already preventing one test from running lots at the start, and instead interleave it with the iterations of other tests? What is the value of having dvsim spend a constant fraction of time on each test? It only ends up stretching further out the tests that are run with fewer iterations.  

With your example, if you believe that testing C more would be helpful, you should probably change the weighting!

FYI, the C tests are directed, so they are run with only once :) I am DVing other blocks as well. 
Not sure how the concept of "weightage" applies here. The # of iterations we set for a test has nothing to do with the # of iterations set for other tests. 

> 
> Do you feel strongly that the current code is wrong? If so, I'll change it to your preferred version. Otherwise, I think the current code is better.

Yes, I think I prefer my version for the reason I explained above. Unless, you can convince me that your approach is categorically better!

Reply:
Ok, I think I understand your motivation. I'll change the code to do as you ask.